beacon-chain/sync/initial-sync/round_robin.go

```diff
@@ -244,48 +262,63 @@ func (s *Service) processBlock(
 	return blockReceiver(ctx, blk, blkRoot)
 }
 
+type processedChecker func(context.Context, blocks.ROBlock) bool
+
+func validUnprocessed(ctx context.Context, bwb []BlockWithVerifiedBlobs, headSlot primitives.Slot, isProc processedChecker) ([]BlockWithVerifiedBlobs, error) {
+	// use a pointer to avoid confusing the zero-value with the case where the first element is processed.
+	var processed *int
+	for i := range bwb {
+		b := bwb[i].block
+		if headSlot >= b.Block().Slot() && isProc(ctx, b) {
+			processed = &i
+			continue
+		}
+		if i > 0 {
+			parent := bwb[i-1].block
+			if parent.Root() != b.Block().ParentRoot() {
+				return nil, fmt.Errorf("expected linear block list with parent root of %#x (slot %d) but received %#x (slot %d)",
+					parent, parent.Block().Slot(), b.Block().ParentRoot(), b.Block().Slot())
+			}
+		}
+	}
+	if processed == nil {
+		return bwb, nil
+	}
+	if *processed+1 == len(bwb) {
+		maxIncoming := bwb[len(bwb)-1].block
+		maxRoot := maxIncoming.Root()
+		return nil, fmt.Errorf("headSlot:%d, blockSlot:%d , root %#x:%w", headSlot, maxIncoming.Block().Slot(), maxRoot, errBlockAlreadyProcessed)
+	}
+	return bwb[*processed:], nil
+}
+
 func (s *Service) processBatchedBlocks(ctx context.Context, genesis time.Time,
-	blks []interfaces.ReadOnlySignedBeaconBlock, bFunc batchBlockReceiverFn) error {
-	if len(blks) == 0 {
+	bwb []BlockWithVerifiedBlobs, bFunc batchBlockReceiverFn) error {
```

Comment:
a lot of these changes might be better suited to be directly against develop, as they seem to be bigger refactors rather than something deneb specific

Reply:
as discussed in our call, these changes are downstream of the change to use the `BlockWithVerifiedBlobs` container. When we get deneb-integration merged to develop we can make changes to add ROBlock in more places.